prow/plugins/projectqueries/projectqueries.go

```diff
@@ -0,0 +1,251 @@
+	}
+	return pluginHelp, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.PluginConfig.ProjectQueries, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectQueries []plugins.ProjectQuery, log *logrus.Entry, pe github.PullRequestEvent) error {
+	if len(projectQueries) == 0 {
+		return nil
+	}
+	// Only respond to label add or issue/PR open events
+	if pe.Action != github.PullRequestActionOpened &&
+		pe.Action != github.PullRequestActionReopened &&
+		pe.Action != github.PullRequestActionLabeled {
+		return nil
+	}
+	// Get any ProjectQueries that match this PR
+	matchedProjectQueries, err := getMatchingProjectQueries(gc, projectQueries, pe)
+	if err != nil {
+		return err
+	}
+	// For each ProjectQuery that matches this PR, add this PR to that project
+	for _, matchedProjectQuery := range matchedProjectQueries {
+		err = addPRToProject(gc, matchedProjectQuery, pe)
+		log.WithError(err).Println("Failed to add PR to project")
+	}
+	return nil
+}
+
+func getMatchingProjectQueries(gc githubClient, projectQueries []plugins.ProjectQuery, pe github.PullRequestEvent) ([]plugins.ProjectQuery, error) {
+	var matchedProjectQueries []plugins.ProjectQuery
+	// Don't use GetIssueLabels unless it's required and keep track of whether the labels have been fetched to avoid unneccessary API usage.
+	labelsFetched := false
+	var labels []github.Label
+	for _, projectQuery := range projectQueries {
+		searchQuery := projectQuery.Query
+		if searchQuery.Org != "" && searchQuery.Org != pe.Repo.Owner.Login {
+			continue
+		}
+		if searchQuery.State != "" && searchQuery.State != pe.PullRequest.State {
+			continue
+		}
+		if len(searchQuery.Labels) != 0 {
+			if !labelsFetched {
+				// If labels are not yet fetched then get them as they are now required
+				// GetIssueLabels works for PRs as they are considered issues in the API
+				var err error
+				labels, err = gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
+				if err != nil {
+					return nil, err
+				}
+				labelsFetched = true
+			}
+			if !github.HasLabels(searchQuery.Labels, labels) {
+				continue
+			}
+		}
+		matchedProjectQueries = append(matchedProjectQueries, projectQuery)
+	}
+	return matchedProjectQueries, nil
+}
+
+func addPRToProject(gc githubClient, projectQuery plugins.ProjectQuery, pe github.PullRequestEvent) error {
+	// Try to find the current project by name
+	currentProject := getProjectByName(projectQuery.Name, orgRepoProjects[projectQuery.Org])
+	// If the project could not be found in the local list then update the it
```

Comment:
Why do we need the caching here? It seems like it would be simpler to just make the API call every time, rather than maintain a potentially invalid list.

We have a ghproxy to cache get requests to solve this sort of thing.

Reply:
The local caching of projects was suggested by @stevekuznetsov to reduce API calls. I'm quite new to the project so not sure what the best practise is for this kind of thing.

Reply:
/cc @cjwagner 
thoughts?

I am always confused about when we do, do not care about tokens

Reply:
If you can be sure that cache won't ever serve stale data then definitely keep it.
We don't want to rely on `ghproxy` if we don't have to so that smaller Prow instances don't need to use it and so that when `ghproxy` is unavailable or has an empty/stale cache, Prow can still operate normally. (e.g. after switching bot accounts or tokens).

`ghproxy` greatly improves our token usage, but it is a single point of failure if Prow becomes reliant on it.

Reply:
I'm now considering changing the plugin config to refer to projects only by number, and columns by ID. This removes the need to make all the API calls to convert names to IDs, or the need for a cache which could become invalid.

Project numbers can be easily found in the GitHub UI. An API call will still be needed to get the actual ID from the org/repo and project number but this can be cached and should never become invalid as you can't change a project number.

Column IDs can be found in the URL obtained using "Copy column link" in the "..." menu.

If the ability to use names was still desired I would suggest that it not use any caching, so users can choose between using numbers and IDs with lower API usage, or full names with more API usage.

Reply:
> We don't want to rely on ghproxy if we don't have to

IMO we want to rely on ghproxy whenever we have to, assuming it involves caching. We don't want every plugin to roll its own cache. This is a recipe for bugs and unnecessary complexity.

If we need caching please depend on ghproxy, don't create something new.

The assumption is either prow is managing repos slow enough that caching doesn't matter. Otherwise deploy ghproxy.

> changing the plugin config to refer to projects only by number

This sounds unfriendly to the people using the plugin.

> projects in an org or org/repo aren't required to have unique names

Technically teams also aren't required to be unique, but since this is hard to work with, peribolos is just going to enforce that team names are unique. I would suggest we do the same thing here (aka if someone references foo then find the first foo project in org/repo else org that matches)